bosonprotocol/core-components
# feat: fetch products with variants directly from subgraph

## Description

In order to reduce the queries from the dApp to the Subgraph when feching products, then variants for each of them, then filtering across offers (for instance removing voided ones, or invalid ones, ...)
- subgraph now includes new properties in the ProductV1Product entity:
  - `variants`: array of all variants for the product (a variant is 1 offer and their variations, for instance size:S and color:Red)
  - `notVoidedVariants`: array of all variants for the product that are not voided
  - `allVariantsVoided`: boolean set to true if all variants of the product are voided (useful to filter products at query level)
  - `minValidFromDate`: minimum value of validFromDate for all variants (useful to filter products at query level)
  - `maxValidFromDate`: maximum value of validFromDate for all variants (useful to filter products at query level)
  - `minValidUntilDate`: minimum value of validUntilDate for all variants (useful to filter products at query level)
    `maxValidUntilDate`: maximum value of validUntilDate for all variants (useful to filter products at query level)
- core-sdk now have 2 new methods:
  - `getAllProductsWithVariants()`: returns all products with the variants array for each of them
  - `getAllProductsWithNotVoidedVariants()`: returns all products with the `notVoidedVariants` array for each of them

  These 2 methods accept filtering at product level: including (for instance):
   - `minValidFromDate_lte`/`maxValidUntilDate_gte` to filter only for valid products (= products that contain valid offers). See e2e:test [here](https://github.com/bosonprotocol/core-components/blob/4fe67d1469feed4b578f73e9b729b07a223b7533/e2e/tests/productV1.test.ts#L729).
   - `allVariantsVoided` to filter only for non-voided products (= products that contain at least 1 non-voided offer). See e2e:test [here](https://github.com/bosonprotocol/core-components/blob/4fe67d1469feed4b578f73e9b729b07a223b7533/e2e/tests/productV1.test.ts#L719)

## How to test

New e2e tests have been added to check each of the new coreSDK methods

Here is an example of what the subgraph looks like:
```
{
  "data": {
    "productV1Products": [
      {
        "id": "4d1d2b2b-1022-4c7e-a4a8-e3f8f82b6b27-1",
        "notVoidedVariants": [
          {
            "offer": {
              "id": "3",
              "voided": false
            },
            "variations": [
              {
                "type": "color",
                "option": "red"
              },
              {
                "type": "size",
                "option": "XS"
              }
            ]
          },
          {
            "offer": {
              "id": "4",
              "voided": false
            },
            "variations": [
              {
                "type": "color",
                "option": "blue"
              },
              {
                "type": "size",
                "option": "S"
              }
            ]
          }
        ],
        "variants": [
          {
            "offer": {
              "id": "3",
              "voided": false,
              "validFromDate": "1668088978",
              "validUntilDate": "1669806168"
            }
          },
          {
            "offer": {
              "id": "4",
              "voided": false,
              "validFromDate": "1668085378",
              "validUntilDate": "1669809768"
            }
          }
        ],
        "minValidFromDate": "1668085378",
        "minValidUntilDate": "1669806168",
        "maxValidFromDate": "1668088978",
        "maxValidUntilDate": "1669809768",
        "allVariantsVoided": false
      },
      {
        "id": "bb1f86f6-dfce-4689-a176-083f01aebe16-1",
        "notVoidedVariants": [
          {
            "offer": {
              "id": "1",
              "voided": false
            },
            "variations": [
              {
                "type": "color",
                "option": "red"
              },
              {
                "type": "size",
                "option": "XS"
              }
            ]
          },
          {
            "offer": {
              "id": "2",
              "voided": false
            },
            "variations": [
              {
                "type": "color",
                "option": "blue"
              },
              {
                "type": "size",
                "option": "S"
              }
            ]
          }
        ],
        "variants": [
          {
            "offer": {
              "id": "1",
              "voided": false,
              "validFromDate": "1668078179",
              "validUntilDate": "1669809768"
            }
          },
          {
            "offer": {
              "id": "2",
              "voided": false,
              "validFromDate": "1668081778",
              "validUntilDate": "1669806168"
            }
          }
        ],
        "minValidFromDate": "1668078179",
        "minValidUntilDate": "1669806168",
        "maxValidFromDate": "1668081778",
        "maxValidUntilDate": "1669809768",
        "allVariantsVoided": false
      }
    ]
  }
}
```
